core/src/main/java/io/questdb/cairo/sql/async/PageFrameReduceJob.java

```diff
@@ -0,0 +1,177 @@
+     *
+     * @param queue  page frame queue instance
+     * @param subSeq subscriber sequence
+     * @param record instance of record that can be positioned on the frame and each row in that frame
+     * @return inverted value of queue processing status; true if nothing was processed.
+     */
+    public static boolean consumeQueue(
+            RingQueue<PageFrameReduceTask> queue,
+            MCSequence subSeq,
+            PageAddressCacheRecord record,
+            SqlExecutionCircuitBreaker circuitBreaker
+    ) {
+        // loop is required to deal with CAS errors, cursor == -2
+        do {
+            final long cursor = subSeq.next();
+            if (cursor > -1) {
+                final PageFrameReduceTask task = queue.get(cursor);
+                final PageFrameSequence<?> frameSequence = task.getFrameSequence();
+                try {
+                    final AtomicInteger framesReducedCounter = frameSequence.getReduceCounter();
+                    try {
+                        LOG.debug()
+                                .$("reducing [shard=").$(frameSequence.getShard())
+                                .$(", id=").$(frameSequence.getId())
+                                .$(", frameIndex=").$(task.getFrameIndex())
+                                .$(", frameCount=").$(frameSequence.getFrameCount())
+                                .$(", valid=").$(frameSequence.isValid())
+                                .$(", cursor=").$(cursor)
+                                .I$();
+                        if (frameSequence.isValid()) {
+                            // we deliberately hold the queue item because
+                            // processing is daisy-chained. If we were to release item before
+                            // finishing reduction, next step (job) will be processing an incomplete task
+                            if (!circuitBreaker.checkIfTripped(frameSequence.getStartTimeUs(), frameSequence.getCircuitBreakerFd())) {
+                                record.of(frameSequence.getSymbolTableSource(), frameSequence.getPageAddressCache());
+                                record.setFrameIndex(task.getFrameIndex());
+                                assert frameSequence.doneLatch.getCount() == 0;
+                                frameSequence.getReducer().reduce(record, task);
+                            } else {
+                                frameSequence.setValid(false);
+                            }
+                        }
+                    } catch (Throwable e) {
+                        frameSequence.setValid(false);
+                        throw e;
+                    } finally {
+                        framesReducedCounter.incrementAndGet();
+                    }
+                } finally {
+                    subSeq.done(cursor);
+                }
+                return false;
+            } else if (cursor == -1) {
+                // queue is full, we should yield or help
```

Comment:
nit: a small question. Since we're on the consumer side, `-1` should mean an empty queue situation, isn't it?

Reply:
you're right, fixed comment here: https://github.com/questdb/questdb/pull/1732/commits/a75ad02362d53f02c86740447dc3dcad3983bdaa